Cleanup on FileSystem tests and resolve issue with leftover temporary files

ianhays opened the pull request:
-Running the FileSystem tests no longer leaves test directories or files in the running directory. fixes #1412
-Modernized the tests for GetFIleSystemEntries and GetFiles, splitting each test case into a standalone method runnable through xunit. Similar versions of these tests now inherit from each other rather than just being copy-pasted like they were previously. Moved most test cases from using hardcoded file names to instead use GetTestFileName() in FileSystemTest.cs. Tested on Linux and Windows.
-Changed the names of some files and classes to better follow the naming conventions of the other FileSystem tests.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs:
```diff
@@ -3,271 +3,147 @@
+    public void FileNameWithSpaces()
+    {
+        String strTempDir = "             ";
+        Assert.Throws<ArgumentException>(() => GetEntries(strTempDir));
```
Hmm.  If this passes on Unix (and it looks like it will, as the same case existed in the old code you're transforming), I think that means we have a bug in our Unix implementation, as an all-spaces name is actually a valid directory name on Unix, e.g.
```
stoub@SToubLinux1:~$ mkdir "    "
stoub@SToubLinux1:~$ cd "    "
stoub@SToubLinux1:~/    $ mkdir something
stoub@SToubLinux1:~/    $ cd something
stoub@SToubLinux1:~/    /something$ 
```

ianhays replied:
That's interesting! Since that is the case, I'll need to go through the tests again and fix this, as whitespace file/directory name checks are in essentially every test.

stephentoub replied:
If the test is passing now, I'd suggest leaving the tests as you have them, opening an issue to track the potential product bug, and adding a new test that's PlatformSpecific.AnyUnix that validates spaces can be used in this manner... you'll need to mark it as ActiveIssue with the issue you opened, as it would currently be failing.  Then after you get these tests merged, you can go back and investigate/fix the problem, and update the then failing tests accordingly.  Just a suggestion.

ianhays replied:
Will do!

ianhays replied:
#2205

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str.cs:
```diff
@@ -3,271 +3,147 @@
+        testDir.CreateSubdirectory("TestDir1");
+        testDir.CreateSubdirectory("TestDir2");
+        testDir.CreateSubdirectory("TestDir3");
+        using (new FileInfo(Path.Combine(TestDirectory, "TestFile1")).Create())
```
Rather than:
```C#
new FileInfo(...).Create()
```
seems like you could simplify slightly to:
```C#
File.Create(...);
```

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs:
```diff
@@ -3,503 +3,171 @@
+    public void ValidSearchPattern()
+    {
+        String strTempDir = "a..b abc..d";
+        GetEntries(TestDirectory, strTempDir); //Should not throw
```
Should this also Assert.Empty?

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs:
```diff
@@ -3,503 +3,171 @@
-            //	Directory.Delete(dirName, true);
+                Assert.Contains(Path.Combine(TestDirectory, "aaabbcc"), strArr);
+            else
+                Assert.DoesNotContain(Path.Combine(TestDirectory, "aaabbcc"), strArr);
```
I think this else case is going to fail on OS X, which like Windows does case-insensitive comparisons by default.  Unfortunately we don't run these tests yet in CI on OS X (we do on Windows and Linux).  Can you test locally?  The condition may need to be updated to have OS X go down the if block.

ianhays replied:
I'll need to wrangle myself up a mac; I don't currently have one available to use. 

The case sensitivity check was frequently brought up in the tests, so again I'll need to go through them and pull out the OS-specific parts. I'll make a new issue for that and mark it Xplat and work on it as I move forward with redoing the worst of the FileSystem tests.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str_so.cs:
```diff
@@ -0,0 +1,73 @@
+    public virtual String[] GetEntries(String dirName, String searchPattern, SearchOption option)
+    {
+        return Directory.GetFileSystemEntries(dirName, searchPattern, option);
+    }
```
Your structuring of these tests is clever. :smile:

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFiles_str_Str.cs:
```diff
@@ -11,454 +12,274 @@
+        // '<' and '>' are valid Unix filename chars
+        if (Interop.IsWindows)
+        {
+            Assert.Throws<ArgumentException>(() => GetFiles(".", "<>"));
```
I think elsewhere where we're checking invalid characters, we have a Unix check as well, using '\0' instead of '>'.  Should we do the same here?

ianhays replied:
Yes, it is happening below in the switch statement - it goes to default and throws an ArgumentException.

Side question, do you know what the justification is for "GetFiles(TestDirectory,'\n'.ToString())" to return the empty array but "GetFiles(Directory.GetCurrentDirectory(), String.Format("te{0}st", '\n'ToString()))" to throw an exception? The distinction of allowing newlines by themselves but not in a name seems a bit inconsistent, unless there's a use case I'm missing.

stephentoub replied:
> The distinction of allowing newlines by themselves but not in a name seems a bit inconsistent, unless there's a use case I'm missing.

I don't think it's that newlines alone are allowed and disallowed otherwise, but that various whitespace charactes (including '\n') are trimmed from the end of the search string (I don't know what the justification is for such trimming, but it's been the behavior for a long time).  So  "\n" and "something\n" are both valid, but "some\nthing" is not: the string is trimmed and then checked for invalid characters.

ianhays commented:
Fixing the temp file/directories issue greatly furthered my desire to redo these FileSystem tests. It seemed better to keep this PR small (since it's my first to the dotnet repo) and fix the issue linked for now before anything else though. I'm going to continue restructuring the test classes to be more like GetFiles and GetFileSystemEntries after this PR goes through, as they desperately need it.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFiles_str_Str.cs:
```diff
@@ -11,454 +12,274 @@
+    [ActiveIssue(846, PlatformID.AnyUnix)]
+    public void CharacterTests()
+    {
+        // TODO: [ActiveIssue(846)] Globalization on Unix
```
Now that we actually have the ActiveIssue on this, the comment for the same thing can be removed.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFiles_str_Str.cs:
```diff
@@ -11,454 +12,274 @@
+            GetFiles(".", ((Char)i).ToString());
+        }
 
 #if DESKTOP
```
Do we ever set this?

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFiles_str_str_so.cs:
```diff
@@ -8,642 +8,321 @@
-                            Console.WriteLine(fileName);
-                    }
-                }
+                if (Eval(list.Contains(files[i]), "Err_287kkm! unexpected file found: {0}", files[i]))
```
With these tests having been overhauled, do we still need these identifiers (e.g. "Err_415mbz!")?

ianhays replied:
If they do have some value, I couldn't find it :) They should be assertions. They're holdovers from before xunit was being used for testing.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFiles_str_str_so.cs:
```diff
@@ -8,642 +8,321 @@
-            /* Scenario disabled when porting because it modifies the filesystem outside of the test's working directory
-            try
+    [Fact]
+    [ActiveIssue(0, PlatformID.Any)]
```
Do we know why this was previously commented out?  If it's just a bad test, now would be a good time to remove it.  If it has value and needs fixing, it'd be good to fix it.  If that's too much work for now, we could tag it with a real issue number rather than 0.

ianhays replied:
I think it's a holdover; the comment says the purpose is to "Ensure that the path contains subdirectories and call this API and ensure that only the top directory files are returned" but when it gets files it uses SearchOption.AllDirectories, which is directly contradicting the point of the test. I don't think it has value since I check subdirectory exclusion/inclusion elsewhere.

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFiles_str_str_so.cs:
```diff
@@ -8,642 +8,321 @@
+        String dirName;
+        String[] expectedFiles;
+        String[] files;
+        List<String> list;
```
These variables don't all need to be declared at the top of the function (same goes for other tests where this is being done).

stephentoub commented on src/System.IO.FileSystem/tests/Directory/GetFiles_str_str_so.cs:
```diff
@@ -8,642 +8,321 @@
+
     //Checks for error
-    private static bool Eval(bool expression, String msg, params Object[] values)
+    protected static bool Eval(bool expression, String msg, params Object[] values)
```
Are these Eval methods just legacy from before we had Assert.Equal?  If so, could we remove them and just use Assert directly?

stephentoub commented on src/System.IO.FileSystem/tests/File/Copy_str_str_b.cs:
```diff
@@ -290,31 +289,31 @@ public static void runTest()
+                        fil1.Delete();
+                        File.Delete("\\TestFile.tmp");
+            #endif
+            */
```
Should this be deleted, or fixed?

ianhays replied:
I haven't looked at the File classes yet beyond fixing the temporary files issue, but at a glance it looks like it should be removed. I think the code formatter made my changes in this section.

stephentoub commented:
Overall, nice cleanup!  The tests are much clearer now, and it's always nice to see so much code gook removed :smile:

In addition to verifying that everything passes on Windows, Linux, and OSX, can you also do a code coverage run before and after to help verify we've not regressed in that regard?  I spot checked in a few places, and it seemed like you did a good job keeping the existing validation, but it'd be good to double check.

ianhays commented:
Thanks @stephentoub! There's still so much to be done in these FileSystem classes! It was hard to leave some of that stuff as is - it hurt my inner designer - but I wanted to get a PR out to fix the persistent temp files issue ASAP. I ran a coverage report after my major changes and didn't have any regression in coverage, but I didn't do one before PR submission so I'll check again.

My next steps are:
* Get a mac and test on it
* Finish restructuring and splitting the tests into separate [Fact] decorated methods
* Remove calls to the unnecessary "Eval" methods
* Get rid of all the debugging strings
* Replace calls to "CheckErrors" with Asserts
* Refactor the "Scenario#" tests and determine which should stay/be removed/moved elsewhere
* Remove calls to Console.WriteLine
* Either modify the CommonUtils.ManageFileSystem class or remove the calls to it. As of now it uses random file names and directory structures which really should be changed to something repeatable and structured.

stephentoub commented:
db, test this please

(failed due to git repo cloning issues earlier in the day)

stephentoub commented:
@dotnet-bot test this please

stephentoub commented:
Looks like there's a file issue that's causing compilation to fail:
```
error CS2001: Source file 'DirectoryInfo\DirectoryInfo_EnumerableAPIs.cs' could not be found [d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.IO.FileSystem\tests\System.IO.FileSystem.Tests.csproj]
```

ianhays commented:
I see. It's failing because I renamed two of the files to match the format of all of the other FileSystem test files. 

I renamed File_EnumerableAPIs.cs to EnumerableAPIs.cs in the "File" test folder as well as DirectoryInfo_EnumerableAPIs.cs to EnumerableAPIs.cs in the "DirectoryInfo" test folder.

ianhays commented:
PR updated! All tests for the modified files are passing, and it should be ready to merge now. Any additional feedback before I go through with the changes?

stephentoub commented:
Thanks, @ianhays.  I left a few more comments.  Most are nits, though the .csproj one should probably be addressed.  Once you've done that, LGTM.  Could you squash before merging?

ianhays commented:
Squashed it down to two commits: one from before PR feedback and one from after PR feedback. I also removed the unnecessary additions to the csproj. I'll go ahead and merge it after the Jenkins build just to be safe.
